dotcom-rendering/src/web/components/App.tsx

```diff
@@ -1273,15 +1280,17 @@ export const App = ({ CAPI, NAV, ophanRecord }: Props) => {
 					/>
 				</Lazy>
 			</Portal>
-			<Portal rootId="bottom-banner">
-				<StickyBottomBanner
-					isSignedIn={isSignedIn}
-					asyncCountryCode={asyncCountryCode}
-					CAPI={CAPI}
-					brazeMessages={brazeMessages}
-					isPreview={!!CAPI.isPreview}
-				/>
-			</Portal>
+			{articleCountIsReady && (
```

Comment:
I wonder should the check for `articleCountIsReady` be pushed down to the underlying components which require it, rather than blocking everything downstream of `StickyBottomBanner`? It's probably not a big deal performance wise but I think right now this will block both the Braze Banner and the CMP banner from rendering? It might also signal the intent more accurately? What do you reckon?

Reply:
The issue is that articleCount is used to decide _if_ a banner should be displayed. I'm not sure the message picker can handle this

Reply:
I chatted to @tomrf1 offline, but I'd started replying to this comment so leaving it here for posterity!

Hey sorry for the delay in replying I've been out for a few days. Ah yep I see what you mean, because the `canShow` of the message picker only runs once. If you did want to work around that I think the way we've handled that elsewhere is to create a `Promise` which represents the eventual completion of `articleCountIsReady` and then await that in the `canShow` of the relevant banner. I think that's how `asyncCountryCode` in `App.tsx` is used. Let me know if I can help at all with this, or if any of this doesn't make sense!

Reply:
If you pass a promise as a prop and then conditionally render based on that doesn't this kind of step outside the declarative mindset of react?

I guess what we'd be doing is using plain javascript to manage state, which is fine, but when we have both react state _plus_ js state don't we then have the problem where we need to manage re-renders manually, potentially creating race conditions? Eg. If I pass in a promise then every component using it will need to setup a useEffect and create some state so that it can trigger a re-render for when the promise resolves. But if you pass in normal react state then that's all built in for free. The component will re-render when the state changes and then you use composition to control things.

So yes, it's definitely true that the `{articleCountIsReady && <MyThing />}` approach blocks _everything_. But you can also `<MyThing articleCountIsReady={articleCountIsReady} />` and then conditionally block locally so that only the think you want to be affected gets re-rendered. This is what React is good at.

Reply:
> If you pass a promise as a prop and then conditionally render based on that doesn't this kind of step outside the declarative mindset of react?

Hmm good question, I'm not sure. Passing a promise directly to a component doesn't feel very different than passing in a callback/handler function which returns a promise which is a fairly common pattern. For example a button with an `onClick` which is provided by the parent component which does something async (a fetch say) and then updates local component state based on the outcome.

> If I pass in a promise then every component using it will need to setup a useEffect and create some state so that it can trigger a re-render for when the promise resolves. But if you pass in normal react state then that's all built in for free.

Also a good point! But eventually you need to deal with that async-ness somewhere in your component hierarchy don't you? I.e. you can't ignore it completely.

In general I agree though that keeping things simple using core React principles is a good thing to aim for. Perhaps the message picker abstraction isn't quite right currently. I think part of the thing here is that we only ever want to show one message per page, i.e. once a banner has been selected to be shown on the current page view, it shouldn't then get superseded by another banner at a later point in time.

Reply:
It's a tricky one indeed!

_Ideally_, all the variables used to decide if a thing should appear would be pure props and then the rendering logic would be simple if statements. I appreciate though that there's more going on with banners, esp. around the decisions on which one to show.

Reply:
Yes, it's not very react-y. But the message picker needs to kick off several concurrent `canShow` Promises, and we don't want to tie each message to article count.

An alternative approach would be to run the message picker logic in a useEffect in App.tsx, and pass the chosen message into the StickyBottomBanner component for rendering (and the same for SlotBodyEnd).

Reply:
> An alternative approach would be to run the message picker logic in a useEffect in App.tsx, and pass the chosen message into the StickyBottomBanner component for rendering (and the same for SlotBodyEnd).

That's an interesting idea. I've no idea how feasible this would be (I think @tjmw would know this better than I) but I like the idea of hoisting the logic around what to display up - I feel like this would help understand things better. At some point this decision logic became relevant to other parts of the page (the SignInGate specifically) so having it happen at a more centralised location is the logical solution to how it could be shared and then this opens up a route for having simpler, more react-y components later.

These are all just thoughts and I don't want to block this PR but I do think its great to have this discussion nonetheless!

Reply:
That approach is basically what happens within `StickyBottomBanner` currently (except we use `useOnce` instead of `useEffect`, as we only want to trigger the picking logic/show a banner message once). So I'm sure what Oliver describes is possible.

Definitely agree that this is all outside of the scope of this PR though! Since it sounds like we've all got ideas on how this should work, probably coming at it from slightly different angles, I'd be up for mobbing on this together at some point if others are keen, to explore/sketch out ideas.

Reply:
I've update the PR to use promises. It might be 'good enough' for merging for now, but I think it would be great to mob at some point to maybe figure out a neater solution